numpy/ma/extras.py

```diff
@@ -708,29 +708,38 @@ def _median(a, axis=None, out=None, overwrite_input=False):
             asorted = a
     else:
         asorted = sort(a, axis=axis)
+
     if axis is None:
         axis = 0
     elif axis < 0:
         axis += a.ndim
+    if axis < 0 or axis >= a.ndim:
+        raise ValueError("Axis value out of bounds")
 
     if asorted.ndim == 1:
         idx, odd = divmod(count(asorted), 2)
-        return asorted[idx - (not odd) : idx + 1].mean()
+        return asorted[idx + odd - 1 : idx + 1].mean()
 
-    counts = asorted.shape[axis] - (asorted.mask).sum(axis=axis)
+    if asorted.mask is nomask:
+        counts = asorted.shape[axis]
+    else:
+        counts = asorted.shape[axis] - (asorted.mask).sum(axis=axis)
     h = counts // 2
+
     # create indexing mesh grid for all but reduced axis
     axes_grid = [np.arange(x) for i, x in enumerate(asorted.shape)
                  if i != axis]
     ind = np.meshgrid(*axes_grid, sparse=True, indexing='ij')
+
     # insert indices of low and high median
     ind.insert(axis, h - 1)
     low = asorted[tuple(ind)]
     low._sharedmask = False
     ind[axis] = h
     high = asorted[tuple(ind)]
+
     # duplicate high if odd number of elements so mean does nothing
-    odd = counts % 2 == 1
+    odd = counts % 2
```

Comment:
This is surely not quite right: the idea here is to get a boolean array for all elements that are odd. But sometimes apparently one gets back a scalar.  Investigating...